Mirrored domain-level permissions

orangejenny opened the pull request:
##### SUMMARY
Rework of https://github.com/dimagi/commcare-hq/pull/27449 There's a good deal of code overlap, but probably easiest to review by commit. Figured the cleaner commit set would make the duplicate review worth it.

##### FEATURE FLAG
This isn't flagged, but it's only visible to domains that have mirrored permissions set up, which requires a dev to configure.

##### RISK ASSESSMENT / QA PLAN
Not sure yet, planning to deploy tomorrow. https://github.com/dimagi/commcare-hq/pull/27449 did pass QA.

##### PRODUCT DESCRIPTION
https://trello.com/c/wRKqEipZ/23-enterprise-web-users-want-to-automatically-have-the-same-level-of-access-across-all-counties

This adds the concept of a domain (the source) allowing one or more other domains (the mirrors) to share its permission. Web users in the source domain are automatically granted their same set of permissions in all of the mirror domains.

This is handled similar to superuser access: these users will be able to log into and modify the linked domains, but they are not considered members of the linked domain, so they will not show up in places like the linked domains' lists of web users.

If a user has a role in both the source and mirror domain, the mirror domain's role is used.

This is no longer related to linked domains in any way.

Mirror domains are listed on the domain select page, labeled "Enterprise":
<img width="1108" alt="81621742-307fb800-93bd-11ea-9209-52cf9cbda06c" src="https://user-images.githubusercontent.com/1486591/82089325-6b396700-96c1-11ea-8cc8-87a85605e9f0.png">

A link to the domain select page is added to the project dropdown, also labeled "Enterprise":
<img width="301" alt="81621674-0201dd00-93bd-11ea-965d-1c1f611808e6" src="https://user-images.githubusercontent.com/1486591/82089286-5957c400-96c1-11ea-9a09-8bd0f8e325f2.png">

End users can't change the configuration, but from the source domain you can view the list of mirror domains:
![Screen Shot 2020-05-15 at 3 24 41 PM](https://user-images.githubusercontent.com/1486591/82088778-5d371680-96c0-11ea-97a2-9b950f430e32.png)

orangejenny committed:
Moved invitations above projects on domain selection page

Prevent this section from getting lost, as the new section may contain dozens of domains.

orangejenny committed:
Added DomainPermissionsMirrorView

orangejenny committed:
Added DomainPermissionsMirror

orangejenny committed:
DRYed up link generation for dropdown and domain select page

orangejenny committed:
Added mirror domains to select page and link to dropdown

orangejenny committed:
Added allow_mirroring flag to is_member_of

orangejenny committed:
Added allow_mirroring flag to get_domain_membership

orangejenny committed:
Added mirror domains to checking for requring secure session

orangejenny committed:
Added 'Enterprise User' label for gear dropdown

orangejenny committed:
Added new models to domain cleanup

orangejenny committed:
Added tests

orangejenny committed:
Added admin site

calellowitz commented on corehq/apps/domain/views/base.py:
```diff
@@ -65,6 +69,22 @@ def select(request, do_not_redirect=False, next_view=None):
+
+@quickcache(['couch_user.username'])
+def get_domain_dropdown_links(couch_user, view_name="domain_homepage"):
+    domain_objects = set(Domain.active_for_user(couch_user))
```
why is this a `set`

orangejenny replied:
It's left over from an earlier iteration. Reworked this code in 0d7c26819078320dbe8f11d6cae15282270a8d81

calellowitz commented on corehq/apps/users/models.py:
```diff
@@ -588,16 +588,20 @@ class _AuthorizableMixin(IsMemberOfMixin):
-                raise self.Inconsistent("Domain '%s' is in domain but not in domain_memberships" % domain)
+            if not domain_membership:
+                if domain in self.domains:
+                    raise self.Inconsistent("Domain '%s' is in domain but not in domain_memberships" % domain)
```
nit: this sentence is really confusing. what does `Domain is in domain` mean?

orangejenny replied:
`MultiMembershipMixin` has a `domains` property and a `domains_membership` property that are apparently supposed to stay in sync. I think we mostly use `domains_membership` and don't know exactly what the `domains` list is for.

calellowitz replied:
does that mean the second `domain` should be plural? That is clearly a minor thing, but it does seem like a typo that makes this more confusing.

calellowitz commented on corehq/apps/users/models.py:
```diff
@@ -465,6 +465,30 @@ class DomainMembershipError(Exception):
 
+class DomainPermissionsMirrorSource(models.Model):
+    # Name of the controlling domain
+    name = models.CharField(max_length=126, db_index=True, unique=True)
```
Can this be a column in the other table? Especially since it looks like neither table is ever called without joining the two

orangejenny replied:
https://github.com/dimagi/commcare-hq/pull/27510/commits/93eb46610c761b27345fb239b935fb564a9f617d

calellowitz reviewed (approved):
Nothing blocking, but I do prefer the one model approach if possible

orangejenny committed:
Limited configuration view to superusers

orangejenny committed:
Updated UI

- On domain select page, combined all projects into one list
- Updated link text and icon in dropdown
- Limit dropdown to 20 items total, so support team also gets the 'All Projects' link
- Fixed bug where invitations weren't showing on domain select page

orangejenny committed:
Updated model definition

calellowitz commented on corehq/apps/domain/views/base.py:
```diff
@@ -70,29 +75,34 @@ def select(request, do_not_redirect=False, next_view=None):
-        _domain_objects_to_links(mirror_domain_objects_by_name.values()),
-    )
+def _domains_to_links(domain_objects, view_name):
+    return sorted([DomainDropdownLink(
```
nit: This is sorted after it gets returned as well
